src/TLabel.cpp

```diff
@@ -114,3 +146,60 @@ void TLabel::enterEvent(QEvent* event)
     }
     QWidget::enterEvent(event);
 }
+
+bool TLabel::forwardEventToMapper(QEvent* event)
+{
+    if (event->type() == QEvent::MouseButtonPress || event->type() == QEvent::MouseButtonRelease || event->type() == QEvent::MouseMove)
+    {
+        QMouseEvent* mouseEvent = static_cast<QMouseEvent*>(event);
+        QWidget* qw = qApp->widgetAt(mouseEvent->globalPos());
+
+        if (qw && this->parentWidget()->findChild<QWidget*>(QStringLiteral("mapper"))->isAncestorOf(qw))
+        {
+            QMouseEvent newEvent(mouseEvent->type(), qw->mapFromGlobal(mouseEvent->globalPos()), mouseEvent->button(), mouseEvent->buttons(), mouseEvent->modifiers());
+            qApp->sendEvent(qw, &newEvent);
+            return true;
+
+        }
+        return false;
+
+    }
+    else if (event->type() == QEvent::Enter || event->type() == QEvent::Leave)
+    {
+        QWidget* qw = qApp->widgetAt(QCursor::pos());
+
+        if (qw && this->parentWidget()->findChild<QWidget*>(QStringLiteral("mapper"))->isAncestorOf(qw))
+        {
+            QEvent newEvent(event->type());
+            qApp->sendEvent(qw, &newEvent);
+            return true;
+
+        }
+        return false;
+    }
+    else if (event->type() == QEvent::Wheel)
+    {
+        QWheelEvent* wheelEvent = static_cast<QWheelEvent*>(event);
+        QWidget* qw = qApp->widgetAt(wheelEvent->globalPos());
+
+        if (qw && this->parentWidget()->findChild<QWidget*>(QStringLiteral("mapper"))->isAncestorOf(qw))
+        {
+            QWheelEvent newEvent(
+                        qw->mapFromGlobal(wheelEvent->globalPos()),
+                        wheelEvent->globalPos(),
+                        wheelEvent->pixelDelta(),
+                        wheelEvent->angleDelta(),
+                        wheelEvent->angleDelta().y()/8,
+                        Qt::Vertical,
```

Comment:
Not sure I am happy about this and the previous line - I happen to have a mouse that has two scroll wheels and the second one (when I fix the production kernel which does not incorporate such a [patch](https://patchwork.kernel.org/patch/9604835/) yet) will produce horizontal events and I suspect that will confuse this code.

The 2D Mapper already handles the `QWheelEvent::angleDelta()` data though the 3D one does not - if you tweak `void GLWidget::wheelEvent(QWheelEvent* e)` to work from `e->angleDelta().y()` instead of `e->delta()` {a divisor may be needed in that method as well} - you can use zero on the line before this and avoid _horizontal_ wheel movements confusing things, IMHO...

Reply:
This was something I was pretty shaky on. I _think_ that the ```wheelEvent->angleDelta().y()/8``` and ``` Qt::Vertical``` arguments will only be used if the receiving object doesn't support 2D scrolling. That's based on the following line in the QWheelEvent documentation.

> For backwards compatibility, the event can also hold monodirectional wheel event data: qt4Delta specifies the rotation, and qt4Orientation the direction.

Since QWheelEvent offers no constructors without those arguments, I'm assuming they just get ignored if the receiving object is able to handle the 2D scroll. Again, though, I'm very shaky on this, and I'm not sure how I could simulate the horizontal scroll to test things out...

Reply:
@SlySven Try it out, see if the code gets confused, and let us know!